modules/edge-identity/src/identity.rs

```diff
@@ -259,89 +247,27 @@ decl_module! {
 		/// Remove a claim as a claims issuer. Ensures that the sender is an active
 		/// claims issuer. Ensures that the sender has issued a claim over the
 		/// identity provided to the module.
-		pub fn remove_claim(origin, identity_hash: T::Hash) -> Result {
+		pub fn remove_claim(origin, identity_hash: T::Hash, claim_index: usize) -> Result {
 			let _sender = ensure_signed(origin)?;
 
-			let issuers: Vec<T::AccountId> = Self::claims_issuers();
-			ensure!(issuers.iter().any(|id| id == &_sender), "Invalid claims issuer");
 			ensure!(<IdentityOf<T>>::exists(identity_hash), "Invalid identity record");
 
 			let mut claims = Self::claims(identity_hash);
-			ensure!(claims.iter().any(|claim| claim.0 == _sender.clone()), "No existing claim under issuer");
+			ensure!(claims[claim_index].0 == _sender.clone(), "No existing claim under issuer");
 
-			let index = claims.iter().position(|claim| claim.0 == _sender.clone()).unwrap();
-			claims.remove(index);
+			claims.remove(claim_index);
 			<Claims<T>>::insert(identity_hash, claims);
 
 			Ok(())
 		}
 
-		/// Check all pending identities for expiration when each block is
-		/// finalised. Once an identity expires, it is deleted from storage.
-		fn on_finalise(n: T::BlockNumber) {
-			let (expired, valid): (Vec<_>, _) = <IdentitiesPending<T>>::get()
-				.into_iter()
-				.partition(|(_, expiration)| n >= *expiration);
-
-			expired.into_iter().for_each(move |(exp_hash, _)| {
-				<Identities<T>>::mutate(|idents| idents.retain(|hash| hash != &exp_hash));
-				<IdentityOf<T>>::remove(exp_hash);
-				Self::deposit_event(RawEvent::Expired(exp_hash))
-			});
-			<IdentitiesPending<T>>::put(valid);
-		}
-	}
-}
+		pub fn remove_expired_identity(_origin, identity_hash: T::Hash) -> Result {
```

Comment:
As with `finalize_verification`, who do we expect will invoke this function and actually perform the removals?

Reply:
We would have to have some kind of cron job doing this. `on_finalise` can be described as some sort of on-chain cron job, so we can just move that computation off-chain.

Reply:
As discussed with @drewstone i brought back `on_finalise`, to avoid creating additional components for maintaining the system.

Under assumption that `on_finalise` is free, i pushed more logic into it. So, removing a a verified identity from pending array is moved from `verify` to `on_finalise`.